File: app/services/ai/theme_identifier.rb
# frozen_string_literal: true

module Ai
  class ThemeIdentifier < BaseAnalyzer
    SYSTEM_PROMPT = <<~PROMPT
      You are an expert at identifying patterns and themes across product insights.
      Analyze the insights and group them into high-level themes that represent
      common patterns, problem areas, or opportunity spaces.

      For each theme:
      1. Create a clear, memorable name
      2. Write a description explaining what this theme encompasses
      3. Assign a priority score (0-100) based on combined severity and user impact
      4. Estimate total affected users across all insights in this theme
      5. List which insights (by index, 0-based) belong to this theme

      An insight can belong to multiple themes if relevant.

      Respond in JSON format:
      {
        "themes": [
          {
            "name": "Theme name",
            "description": "What this theme represents",
            "priority_score": 85,
            "affected_users_estimate": 500,
            "insight_indices": [0, 2, 5, 7]
          }
        ]
      }

      Guidelines:
      - Create 3-7 themes that meaningfully group the insights
      - Theme names should be concise and memorable (2-4 words)
      - Higher priority for themes with many critical/major insights
      - Don't create themes for single insights unless they're significant
      - Look for both problem themes and opportunity themes
    PROMPT

    def identify(insights)
      return { themes: [], created: 0 } if insights.empty?

      insight_array = insights.to_a
      result = analyze_insights(insight_array)

      if result[:error]
        Rails.logger.error("Theme identification failed: #{result[:error]}")
        return { themes: [], created: 0, error: result[:error] }
      end

      created = create_themes_from_result(result, insight_array)

      { themes: created, created: created.count }
    end

    def reanalyze_all
      insights = Insight.actionable.includes(:themes)
      identify(insights)
    end

    private

    def analyze_insights(insights)
      prompt = build_insights_prompt(insights)
      call_claude(prompt, system_prompt: SYSTEM_PROMPT)
    end

    def build_insights_prompt(insights)
      parts = ["Analyze the following #{insights.count} product insights and identify themes:\n"]

      insights.each_with_index do |insight, index|
        parts << "--- Insight ##{index} ---"
        parts << "Title: #{insight.title}"
        parts << "Description: #{insight.description}"
        parts << "Type: #{insight.insight_type}"
        parts << "Severity: #{insight.severity}"
        parts << "Affected Users: #{insight.affected_users_count}"
        parts << "Feedback Count: #{insight.feedback_count}"
        parts << ""
      end

      parts.join("\n")
    end

    def create_themes_from_result(result, insights)
      created_themes = []

      themes_data = result[:themes] || []
      themes_data.each do |theme_data|
        theme = find_or_create_theme(theme_data)
        next unless theme

        link_insights_to_theme(theme, theme_data, insights)
        theme.update_insight_count!
        theme.update!(analyzed_at: Time.current)

        created_themes << theme
      end

      created_themes
    end

    def find_or_create_theme(data)
      existing = Theme.find_by("LOWER(name) = ?", data[:name].downcase)
      return update_theme(existing, data) if existing

      Theme.create!(
        name: data[:name],
        description: data[:description],
        priority_score: data[:priority_score] || 0,
        affected_users_estimate: data[:affected_users_estimate] || 0
      )
    rescue ActiveRecord::RecordInvalid => e
      Rails.logger.error("Failed to create theme: #{e.message}")
      nil
    end

    def update_theme(theme, data)
      theme.update!(
        description: data[:description],
        priority_score: [theme.priority_score, data[:priority_score] || 0].max,
        affected_users_estimate: [theme.affected_users_estimate, data[:affected_users_estimate] || 0].max
      )
      theme
    end

    def link_insights_to_theme(theme, data, insights)
      indices = data[:insight_indices] || []
      indices.each do |index|
        insight = insights[index]
        next unless insight

        InsightTheme.find_or_create_by!(
          insight: insight,
          theme: theme
        ) do |it|
          it.relevance_score = calculate_relevance(index, indices)
        end
      end
    end

    def calculate_relevance(index, all_indices)
      return 1.0 if all_indices.count == 1
      (1.0 - (all_indices.index(index).to_f / all_indices.count)).round(2)
    end
  end
end

File: apps/worker/src/ai/theme-identifier.ts
/**
 * AI theme identification — ports Rails ThemeIdentifier.
 * Groups insights into high-level themes.
 */
import { eq, desc, sql } from "drizzle-orm";
import type { Database } from "@customer-pulse/db/client";
import { insights, themes, insightThemes } from "@customer-pulse/db/client";
import { callClaudeJson, resolveApiKey } from "./call-claude.js";

interface IdentifiedTheme {
  name: string;
  description: string;
  priority_score: number;
  affected_users_estimate: number;
  insight_ids: number[];
  relevance_scores: Record<string, number>;
}

const SYSTEM_PROMPT = `You are a product strategist analyzing customer insights. Group these insights into 3-7 high-level themes that capture the major patterns.

For each theme, return:
- "name": short theme name (< 60 chars)
- "description": 1-2 sentence description of what this theme represents
- "priority_score": 0-100 based on severity and breadth of the underlying insights
- "affected_users_estimate": total estimated users affected across all insights in this theme
- "insight_ids": array of insight IDs that belong to this theme
- "relevance_scores": object mapping insight_id to relevance score (0.0-1.0)

An insight can belong to multiple themes. Higher priority scores should be given to themes with more severe or widespread issues.

Return a JSON array of theme objects.
Respond with ONLY the JSON array, no markdown fences.`;

export async function identifyThemes(
  db: Database,
  projectId: number,
): Promise<number> {
  const apiKey = await resolveApiKey();
  if (!apiKey) return 0;

  const allInsights = await db
    .select({ id: insights.id, title: insights.title, description: insights.description, insightType: insights.insightType, severity: insights.severity, affectedUsersCount: insights.affectedUsersCount })
    .from(insights)
    .where(eq(insights.projectId, projectId))
    .orderBy(desc(insights.createdAt))
    .limit(50);

  if (allInsights.length < 2) return 0;

  const typeNames: Record<number, string> = { 0: "problem", 1: "opportunity", 2: "trend", 3: "risk", 4: "user_need" };
  const severityNames: Record<number, string> = { 0: "informational", 1: "minor", 2: "moderate", 3: "major", 4: "critical" };

  const insightList = allInsights
    .map((i) => `[ID:${i.id}] [${typeNames[i.insightType]}/${severityNames[i.severity]}] ${i.title}: ${i.description.slice(0, 150)}`)
    .join("\n");

  const result = await callClaudeJson<IdentifiedTheme[]>({
    system: SYSTEM_PROMPT,
    user: `Analyze these ${allInsights.length} insights and group them into themes:\n\n${insightList}`,
    maxTokens: 4096,
  });

  if (!result || !Array.isArray(result)) return 0;

  const now = new Date();
  let created = 0;

  for (const item of result) {
    // Find or create theme by name (case-insensitive)
    const [existing] = await db
      .select({ id: themes.id })
      .from(themes)
      .where(sql`lower(${themes.name}) = ${(item.name ?? "").toLowerCase()} AND ${themes.projectId} = ${projectId}`)
      .limit(1);

    let themeId: number;
    if (existing) {
      await db.update(themes).set({
        description: item.description ?? undefined,
        priorityScore: item.priority_score ?? 0,
        affectedUsersEstimate: item.affected_users_estimate ?? 0,
        insightCount: item.insight_ids?.length ?? 0,
        analyzedAt: now,
        updatedAt: now,
      }).where(eq(themes.id, existing.id));
      themeId = existing.id;
    } else {
      const [newTheme] = await db.insert(themes).values({
        projectId,
        name: item.name ?? "Unnamed Theme",
        description: item.description ?? null,
        priorityScore: item.priority_score ?? 0,
        affectedUsersEstimate: item.affected_users_estimate ?? 0,
        insightCount: item.insight_ids?.length ?? 0,
        metadata: {},
        analyzedAt: now,
        createdAt: now,
        updatedAt: now,
      }).returning({ id: themes.id });
      if (!newTheme) continue;
      themeId = newTheme.id;
      created++;
    }

    // Link insights to theme
    if (item.insight_ids) {
      for (const insightId of item.insight_ids) {
        const relevance = item.relevance_scores?.[String(insightId)] ?? 0.5;
        try {
          await db.insert(insightThemes).values({
            insightId,
            themeId,
            relevanceScore: relevance,
            createdAt: now,
            updatedAt: now,
          });
        } catch { /* duplicate — ignore */ }
      }
    }
  }

  return created;
}
